File: android/src/main/kotlin/dev/ae1/local_translation/LocalTranslationPlugin.kt
package dev.ae1.local_translation

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.util.concurrent.Executors

class LocalTranslationPlugin :
    FlutterPlugin,
    ActivityAware,
    LocalTranslationHostApi {
    private var languageDetector: LanguageDetector? = null
    private var onDeviceTranslator: OnDeviceTranslator? = null
    private val executor = Executors.newSingleThreadExecutor()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val context = flutterPluginBinding.applicationContext
        val detector = LanguageDetector(context)
        languageDetector = detector
        onDeviceTranslator = OnDeviceTranslator(context, detector)
        LocalTranslationHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        LocalTranslationHostApi.setUp(binding.binaryMessenger, null)
        onDeviceTranslator?.close()
        languageDetector = null
        onDeviceTranslator = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        onDeviceTranslator?.activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        onDeviceTranslator?.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onDeviceTranslator?.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDeviceTranslator?.activity = null
    }

    override fun isSupported(callback: (Result<Boolean>) -> Unit) {
        executor.execute {
            val translator = onDeviceTranslator
            if (translator == null) {
                callback(Result.success(false))
                return@execute
            }
            callback(runCatching { translator.isSupported() })
        }
    }

    override fun detectLanguage(
        text: String,
        callback: (Result<HostLanguageDetection>) -> Unit,
    ) {
        val detector = languageDetector
        if (detector == null) {
            callback(Result.failure(unsupported()))
            return
        }
        detector.detect(text, callback)
    }

    override fun detectLanguages(
        texts: List<String>,
        callback: (Result<List<HostLanguageDetection>>) -> Unit,
    ) {
        val detector = languageDetector
        if (detector == null) {
            callback(Result.failure(unsupported()))
            return
        }
        detector.detectLanguages(texts, callback)
    }

    override fun translate(
        request: HostTranslateRequest,
        callback: (Result<HostTranslationResult>) -> Unit,
    ) {
        val translator = onDeviceTranslator
        if (translator == null) {
            callback(Result.failure(unsupported()))
            return
        }
        translator.translate(request, callback)
    }

    override fun translateBatch(
        request: HostTranslateBatchRequest,
        callback: (Result<List<HostTranslationResult>>) -> Unit,
    ) {
        val translator = onDeviceTranslator
        if (translator == null) {
            callback(Result.failure(unsupported()))
            return
        }
        translator.translateBatch(request, callback)
    }

    private fun unsupported(): FlutterError {
        return FlutterError(
            "unsupportedPlatform",
            "Translation is not available",
            null,
        )
    }
}

File: ios/local_translation/Sources/local_translation/LocalTranslationPlugin.swift
import Flutter
import UIKit

public class LocalTranslationPlugin: NSObject, FlutterPlugin, LocalTranslationHostApi {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = LocalTranslationPlugin()
    LocalTranslationHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
  }

  func isSupported() throws -> Bool {
    if #available(iOS 18.0, *) {
      #if targetEnvironment(simulator)
        return false
      #else
        return true
      #endif
    }
    return false
  }

  func detectLanguage(
    text: String,
    completion: @escaping (Result<HostLanguageDetection, Error>) -> Void
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      completion(.success(LanguageDetector.detect(text: text)))
    }
  }

  func detectLanguages(
    texts: [String],
    completion: @escaping (Result<[HostLanguageDetection], Error>) -> Void
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      completion(.success(LanguageDetector.detect(texts: texts)))
    }
  }

  func translate(
    request: HostTranslateRequest,
    completion: @escaping (Result<HostTranslationResult, Error>) -> Void
  ) {
    guard let text = request.text, let targetLanguage = request.targetLanguage else {
      completion(.failure(Self.missingArgument("Missing text or target language")))
      return
    }
    #if canImport(Translation)
      if #available(iOS 18.0, *) {
        Task { @MainActor in
          do {
            let value = try await TranslationSessionHost.shared.translate(
              text: text,
              sourceLanguage: request.sourceLanguage,
              targetLanguage: targetLanguage
            )
            completion(.success(value))
          } catch {
            completion(.failure(Self.mapError(error)))
          }
        }
        return
      }
    #endif
    completion(.failure(Self.unsupportedPlatform))
  }

  func translateBatch(
    request: HostTranslateBatchRequest,
    completion: @escaping (Result<[HostTranslationResult], Error>) -> Void
  ) {
    guard let texts = request.texts, let targetLanguage = request.targetLanguage else {
      completion(.failure(Self.missingArgument("Missing texts or target language")))
      return
    }
    #if canImport(Translation)
      if #available(iOS 18.0, *) {
        Task { @MainActor in
          do {
            let value = try await TranslationSessionHost.shared.translateBatch(
              texts: texts,
              sourceLanguage: request.sourceLanguage,
              targetLanguage: targetLanguage
            )
            completion(.success(value))
          } catch {
            completion(.failure(Self.mapError(error)))
          }
        }
        return
      }
    #endif
    completion(.failure(Self.unsupportedPlatform))
  }

  private static let unsupportedPlatform = PigeonError(
    code: "unsupportedPlatform",
    message: "Translation requires iOS 18 or later on a physical device",
    details: nil
  )

  private static func missingArgument(_ message: String) -> PigeonError {
    PigeonError(code: "unknown", message: message, details: nil)
  }

  private static func mapError(_ error: Error) -> PigeonError {
    if let pigeonError = error as? PigeonError {
      return pigeonError
    }
    let message = error.localizedDescription
    let lower = message.lowercased()
    if lower.contains("cancel") || lower.contains("dismiss") {
      return PigeonError(code: "cancelled", message: message, details: nil)
    }
    if lower.contains("unsupported") || lower.contains("language") {
      return PigeonError(code: "unsupportedLanguagePair", message: message, details: nil)
    }
    return PigeonError(code: "unknown", message: message, details: nil)
  }
}
